crates/shared/src/account_balances/simulation.rs

```diff
@@ -86,16 +144,27 @@ struct Simulation {
 #[async_trait::async_trait]
 impl BalanceFetching for Balances {
     async fn get_balances(&self, queries: &[Query]) -> Vec<Result<U256>> {
+        let mut tokens: HashMap<_, _> = Default::default();
+        for query in queries {
+            if query.interactions.is_empty() {
+                tokens
+                    .entry(query.token)
+                    .or_insert_with(|| contracts::ERC20::at(&self.web3, query.token));
+            }
+        }
+
         // TODO(nlordell): Use `Multicall` here to use fewer node round-trips
         let futures = queries
             .iter()
             .map(|query| async {
-                let simulation = self.simulate(query, None).await?;
-                Ok(if simulation.can_transfer {
-                    simulation.effective_balance
+                if query.interactions.is_empty() {
+                    let token = tokens
+                        .get(&query.token)
```

Comment:
By now it's no longer really expensive to create an instance of a contract so let's drop the `tokens` map and instead create every token on demand here to avoid this error handling/complexity.